### PR TITLE
UNR-5530 Gameplay debugger interest setup

### PR DESCRIPTION
**Pull request by improbable-fabian**
#### Description
This PR changes the `InterestFactory` to add an extra query to `GameplayDebuggerCategoryReplicator` actors ensuring that the auth server can see the owning `PlayerController`. Right now, this is a no-op change since the auth server for the player controller and the replicator are always the same one, so interest is maintained through authority. However, once we start decoupling them (see [design doc](https://brevi.link/multiworker-gameplay-debugger-design)) we have to ensure that whatever server gains authority over the replicator doesn't lose track of the owning player controller actor.

#### Release note
N/A, will follow as part of full feature release

#### Tests
Inspector shows the query being successfully added in the multi worker world composition gym:
pcinterest
In this case, entity 3009 is the one representing the player controller.

#### Documentation
N/A, will follow as part of full feature release

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 1]
```diff
@@ -18,6 +18,10 @@
 #include "GameFramework/PlayerController.h"
 #include "UObject/UObjectIterator.h"
 
+#if WITH_UNREAL_DEVELOPER_TOOLS || (!UE_BUILD_SHIPPING && !UE_BUILD_TEST)
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:21]: Should this be WITH_GAMEPLAY_DEBUGGER?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 2]
```diff
@@ -239,6 +243,14 @@ Interest InterestFactory::CreateInterest(AActor* InActor, const FClassInfo& InIn
 		AddClientPlayerControllerActorInterest(ResultInterest, InActor, InInfo);
 	}
 
+#if WITH_UNREAL_DEVELOPER_TOOLS || (!UE_BUILD_SHIPPING && !UE_BUILD_TEST)
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:246]: WITH_GAMEPLAY_DEBUGGER?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 3]
```diff
@@ -239,6 +243,14 @@ Interest InterestFactory::CreateInterest(AActor* InActor, const FClassInfo& InIn
...
+	if (InActor->IsA(AGameplayDebuggerCategoryReplicator::StaticClass()))
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:247]: Rather do the isa/cast conversion twice, this perhaps could be written as:

```
if (AGameplayDebuggerCategoryReplicator* Replicator = Cast<AGameplayDebuggerCategoryReplicator>(InActor))
{
    AddServerGameplayDebuggerCategoryReplicatorActorInterest(ResultInterest, *Replicator);
}

```
This has the addition benefit of being a safe usage of a potentially null InActor pointer, which unfortunately the rest of this function isn't.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 4]
```diff
@@ -269,6 +281,31 @@ void InterestFactory::AddClientPlayerControllerActorInterest(Interest& OutIntere
 	}
 }
 
+#if WITH_UNREAL_DEVELOPER_TOOLS || (!UE_BUILD_SHIPPING && !UE_BUILD_TEST)
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:284]: WITH_GAMEPLAY_DEBUGGER

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 5]
```diff
@@ -269,6 +281,31 @@ void InterestFactory::AddClientPlayerControllerActorInterest(Interest& OutIntere
...
+void InterestFactory::AddServerGameplayDebuggerCategoryReplicatorActorInterest(Interest& OutInterest,
+																			   const AGameplayDebuggerCategoryReplicator& Replicator) const
+{
+	APlayerController* PlayerController = Replicator.GetReplicationOwner();
+	if (PlayerController == nullptr)
+	{
+		return;
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:291]: Worth a warning?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp [hunk 6]
```diff
@@ -269,6 +281,31 @@ void InterestFactory::AddClientPlayerControllerActorInterest(Interest& OutIntere
...
+	}
+
+	const UNetDriver* NetDriver = PlayerController->GetNetDriver();
+	if (NetDriver == nullptr)
+	{
+		return;
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:297]: Ditto

File: SpatialGDK/Source/SpatialGDK/Public/Utils/InterestFactory.h [hunk 7]
```diff
@@ -69,6 +70,11 @@ class SPATIALGDK_API InterestFactory
 
 	// Defined Constraint AND Level Constraint
 	void AddClientPlayerControllerActorInterest(Interest& OutInterest, const AActor* InActor, const FClassInfo& InInfo) const;
+#if WITH_UNREAL_DEVELOPER_TOOLS || (!UE_BUILD_SHIPPING && !UE_BUILD_TEST)
```
Comment (martin-improbable) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/InterestFactory.h:73]: WITH_GAMEPLAY_DEBUGGER

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 1 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:21]: I copied this from `SpatialActorUtils.h`, and I guess I'll change it there as well!

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 2 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:246]: fixed

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 3 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:247]: changed

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 4 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:284]: fixed

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 5 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:291]: added

File: SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp (hunk 6 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp:297]: added

File: SpatialGDK/Source/SpatialGDK/Public/Utils/InterestFactory.h (hunk 7 above)
Comment (improbable-fabian) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/InterestFactory.h:73]: fixed

**Comment by improbable-fabian** [comment]
addressed comments, PTAL

**Review by martin-improbable — APPROVED** [review]

**Pull request merged**